juniper/src/executor/look_ahead_lazy.rs

```diff
@@ -0,0 +1,1561 @@
+                },
+                LookAheadValue::from_input_value(&val.item, &val.span, vars),
+            )
+        })
+    }
+}
+
+impl<'a, S> Default for LookAheadObject<'a, S> {
+    fn default() -> Self {
+        Self {
+            input_object: &[],
+            vars: None,
+        }
+    }
+}
+
+impl<'a, S: ScalarValue> PartialEq for LookAheadObject<'a, S> {
+    fn eq(&self, other: &Self) -> bool {
+        self.iter().eq(other.iter())
+    }
+}
+
+/// An argument passed into the query
+#[derive(Clone, Copy, Debug)]
+pub struct LookAheadArgument<'a, S> {
+    name: &'a Spanning<&'a str>,
+    input_value: &'a Spanning<InputValue<S>>,
+    vars: &'a Variables<S>,
+}
+
+impl<'a, S: ScalarValue> LookAheadArgument<'a, S> {
+    /// The argument's name
+    pub fn name(&self) -> &'a str {
+        self.name.item
+    }
+
+    /// The Span of the argument's name
+    pub fn name_span(&self) -> &'a Span {
+        &self.name.span
+    }
+
+    /// The argument's value
+    pub fn value(&self) -> LookAheadValue<'a, S> {
+        LookAheadValue::from_input_value(
+            &self.input_value.item,
+            &self.input_value.span,
+            Some(self.vars),
+        )
+        .item
+    }
+
+    /// The Span of the argument's value
+    pub fn value_span(&self) -> &'a Span {
+        &self.input_value.span
+    }
+}
+
+/// The arguments passed into a query.
+#[derive(Copy, Clone, Debug)]
+pub struct LookAheadArguments<'a, S> {
+    arguments: &'a Arguments<'a, S>,
+    vars: &'a Variables<S>,
+}
+
+impl<'a, S> LookAheadArguments<'a, S> {
+    /// Returns an iterator over the arguments.
+    pub fn iter(&self) -> impl Iterator<Item = LookAheadArgument<'a, S>> {
+        let vars = self.vars;
+        self.arguments
+            .items
+            .iter()
+            .map(|(name, input_value)| LookAheadArgument {
+                name,
+                input_value,
+                vars,
+            })
+    }
+}
+
+/// The children of a selection.
+#[derive(Clone, Debug)]
+pub struct LookAheadChildren<'a, S: ScalarValue + 'a> {
+    children: Vec<LookAheadSelection<'a, S>>,
+}
+
+impl<'a, S: ScalarValue> Default for LookAheadChildren<'a, S> {
+    fn default() -> Self {
+        Self { children: vec![] }
+    }
+}
+
+impl<'a, S: ScalarValue> LookAheadChildren<'a, S> {
+    /// Returns the number of children present.
+    pub fn len(&self) -> usize {
+        self.children.len()
+    }
+
+    /// Indicates whether the current node has any children.
+    pub fn is_empty(&self) -> bool {
+        self.children.is_empty()
+    }
+
+    /// Returns the child selection for the specified field.
+    ///
+    /// If a child has an alias, it will only match if the alias matches the specified `name`.
+    pub fn select(&self, name: &str) -> Option<LookAheadSelection<'a, S>> {
+        self.children
+            .iter()
+            .find(|child| child.field_name() == name)
+            .cloned()
```

Comment:
I don't see how, unless `SelectionSource` and `LookAheadSelection` have _manual_ implementations of `Clone` and `Copy` (instead of `#[derive]`). The problem with the derive is with the generic parameter `S: ScalarValue`; Copy gets conditionally implemented given `S: Copy`, which is false.

Reply:
@audunhalland yup, `std` derives are awfully dumb when it comes to generics, so providing a manual `Copy` impl is totally OK here.

Reply:
Fixed in https://github.com/graphql-rust/juniper/pull/1212/commits/149e61f3b52226121df03f43f3cd377315368b17.